utils/retrieve-state/src/lib.rs

```diff
@@ -289,50 +337,369 @@ pub fn put_block_with_deploys(
+                println!("state root has no missing descendants.");
+                println!(
+                    "tries downloaded: {downloaded}/{total}",
+                    downloaded = tries_downloaded,
+                    total = total_tries_count,
+                );
+                break;
+            } else {
+                println!(
+                    "state root still has {} missing descendants. {:#?}",
+                    missing.len(),
+                    in_flight_counters,
+                );
+            }
+        }
+    }
+
+    for (_address, peer) in peer_map.drain() {
+        peer.shutdown.send(()).unwrap();
+        peer.join_handle.await.unwrap();
+    }
+
+    Ok(0)
+}
+
+struct Peer {
+    address: SocketAddr,
+    sender: UnboundedSender<PeerMsg>,
+    join_handle: JoinHandle<()>,
+    shutdown: oneshot::Sender<()>,
+}
+
+async fn maybe_construct_peer(
+    base_send: UnboundedSender<Result<PeerMsg, PeerError>>,
+    address: SocketAddr,
+    client: &Client,
+    state_root_hash: Digest,
+) -> Result<Peer, PeerError> {
+    let url = format!("http://{}:{}/rpc", address.ip(), address.port());
+    let base_send = base_send.clone();
+    let client = client.clone();
+    let read_future = get_trie(
+        &client,
+        &url,
+        GetTrieParams {
+            trie_key: state_root_hash,
+        },
+    );
+    let read_or_timeout = tokio::time::timeout(Duration::from_secs(1), read_future).await;
+    match read_or_timeout {
+        Ok(Ok(_)) => Ok(spawn_peer(base_send, address, client)),
+        Ok(Err(error)) => {
+            return Err(PeerError::FailedRpc {
+                peer_address: address,
+                trie_key: state_root_hash,
+                error,
+            })
+        }
+        Err(elapsed) => {
+            return Err(PeerError::TimedOut {
+                peer_address: address,
+                trie_key: state_root_hash,
+                error: elapsed,
+            })
         }
-        if tries_downloaded % 1000 == 0 {
-            println!(
-                "downloaded {} tries in {}ms",
-                tries_downloaded,
-                start.elapsed().as_millis()
-            );
-            start = Instant::now();
+    }
+}
+
+/// Spawns a task to handle requests sent to this peer. This task will query the peer for a given
+/// Trie under a Digest. The returned Digest will be written to LMDB global state.
+fn spawn_peer(
+    base_send: UnboundedSender<Result<PeerMsg, PeerError>>,
+    address: SocketAddr,
+    client: Client,
+) -> Peer {
+    let (sender, peer_recv) = futures_channel::mpsc::unbounded::<PeerMsg>();
+    let peer_recv = peer_recv.map(|msg| Either::Left(msg));
+
+    let (shutdown, shutdown_recv) = futures_channel::oneshot::channel::<()>();
+    let shutdown_recv = shutdown_recv.into_stream().map(|_| Either::Right(()));
+
+    let mut next_peer_msg_or_shutdown_stream = futures::stream::select(peer_recv, shutdown_recv);
```

Comment:
Why not just a single stream, and a `Shutdown` message variant?

Reply:
The idiom here is adapting a oneshot into a stream that is selected over with the response from the other stream. Warp uses this sort of api here https://docs.rs/warp/0.3.2/warp/struct.Server.html#method.serve_incoming_with_graceful_shutdown